### PR TITLE
Rewrite code-block to TypeScript

### LINKED ISSUE #13008: [TS] Rewrite ckeditor5-code-block into TypeScript
Part of #11704.

### PR DESCRIPTION
**Pull request by mremiszewski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Type: Rewrite code-block to TypeScript. Closes #13008.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 1]
```diff
@@ -181,34 +203,41 @@ export default class CodeBlockEditing extends Plugin {
 		// upon next paste, this bare text will not be inserted as a code block, which is not the best UX.
 		// Similarly, when the selection in a single line, the selected content should be an inline code
 		// so it can be pasted later on and retain it's preformatted nature.
-		this.listenTo( model, 'getSelectedContent', ( evt, [ selection ] ) => {
+		this.listenTo<ModelGetSelectedContentEvent>( model, 'getSelectedContent', ( evt, [ selection ] ) => {
 			const anchor = selection.anchor;
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockediting.ts:206]:
```suggestion
			const anchor = selection.anchor!;
```

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 2]
```diff
@@ -181,34 +203,41 @@ export default class CodeBlockEditing extends Plugin {
...
 
-			if ( selection.isCollapsed || !anchor.parent.is( 'element', 'codeBlock' ) || !anchor.hasSameParentAs( selection.focus ) ) {
+			if (
+				!anchor ||
+				selection.isCollapsed ||
+				!anchor.parent.is( 'element', 'codeBlock' ) ||
+				!anchor.hasSameParentAs( selection.focus! )
+			) {
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockediting.ts:213]:
```suggestion
			if (
				selection.isCollapsed ||
				!anchor.parent.is( 'element', 'codeBlock' ) ||
				!anchor.hasSameParentAs( selection.focus! )
			) {
```

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 3]
```diff
@@ -256,28 +285,31 @@ export default class CodeBlockEditing extends Plugin {
...
+ *
+ * // After pressing:
+ * <codeBlock>
+ * "    foo()"                 // Indent of 4 spaces.
+ * <softBreak></softBreak>     // A new soft break created by pressing enter.
+ * "    "[]                    // Retain the indent of 4 spaces.
+ * </codeBlock>
+ * ```
+ */
+
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockediting.ts:305]:
```suggestion
```

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 4]
```diff
@@ -290,29 +322,33 @@ function breakLineOnEnter( editor ) {
 
 		// If the line before being broken in two had some indentation, let's retain it
 		// in the new line.
-		if ( leadingWhiteSpaces ) {
+		if ( leadingWhiteSpaces && modelDoc.selection.anchor ) {
 			writer.insertText( leadingWhiteSpaces, modelDoc.selection.anchor );
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockediting.ts:325]:
```suggestion
		if ( leadingWhiteSpaces ) {
			writer.insertText( leadingWhiteSpaces, modelDoc.selection.anchor! );
```

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 5]
```diff
@@ -290,29 +322,33 @@ function breakLineOnEnter( editor ) {
...
+ * <codeBlock>[]<softBreak></softBreak>foo</codeBlock>
+ *
+ * // After pressing:
+ * <paragraph>[]</paragraph><codeBlock>foo</codeBlock>
+ * ```
+ *
+ * @param isSoftEnter When `true`, enter was pressed along with <kbd>Shift</kbd>.
+ * @returns `true` when selection left the block. `false` if stayed.
+ */
+
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockediting.ts:345]:
```suggestion
```

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 6]
```diff
@@ -346,29 +382,34 @@ function leaveBlockStartOnEnter( editor, isSoftEnter ) {
...
+ *
+ * // After second press:
+ * <codeBlock>foo</codeBlock><paragraph>[]</paragraph>
+ * ```
+ *
+ * @param editor
+ * @param isSoftEnter When `true`, enter was pressed along with <kbd>Shift</kbd>.
+ * @returns `true` when selection left the block. `false` if stayed.
+ */
+
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockediting.ts:403]:
```suggestion
```

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 7]
```diff
@@ -346,29 +382,34 @@ function leaveBlockStartOnEnter( editor, isSoftEnter ) {
...
+ * <codeBlock>foo[]</codeBlock>
+ *
+ * // After first press:
+ * <codeBlock>foo<softBreak></softBreak>[]</codeBlock>
+ *
+ * // After second press:
+ * <codeBlock>foo</codeBlock><paragraph>[]</paragraph>
+ * ```
+ *
+ * @param editor
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockediting.ts:399]:
```suggestion
```

File: packages/ckeditor5-code-block/src/codeblockui.ts [hunk 8]
```diff
@@ -27,21 +28,21 @@ export default class CodeBlockUI extends Plugin {
 	/**
 	 * @inheritDoc
 	 */
-	static get pluginName() {
+	public static get pluginName(): string {
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockui.ts:31]:
```suggestion
	public static get pluginName(): 'CodeBlockUI' {
```

Comment (niegowski) [reply packages/ckeditor5-code-block/src/codeblockui.ts:31]: And add this to `PluginsMap` (all plugins should be registered in `PluginsMap` type).

File: packages/ckeditor5-code-block/src/codeblockui.ts [hunk 9]
```diff
@@ -64,7 +65,7 @@ export default class CodeBlockUI extends Plugin {
 
 			dropdownView.on( 'execute', evt => {
 				editor.execute( 'codeBlock', {
-					language: evt.source._codeBlockLanguage,
+					language: ( evt.source as { _codeBlockLanguage: string } )._codeBlockLanguage,
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblockui.ts:68]: Maybe we should consider adding typings for this `EventInfo`?

File: packages/ckeditor5-code-block/src/converters.ts [hunk 10]
```diff
@@ -7,31 +7,42 @@
...
+ * ```html
+ * <pre><code class="language-javascript">foo();\nbar();</code></pre>
+ * ```
  *
- * @param {module:engine/model/model~Model} model
- * @param {Array.<module:code-block/codeblock~CodeBlockLanguageDefinition>} languageDefs The normalized language
- * configuration passed to the feature.
- * @param {Boolean} [useLabels=false] When `true`, the `<pre>` element will get a `data-language` attribute with a
+ * @param languageDefs The normalized language configuration passed to the feature.
+ * @param [useLabels=false] When `true`, the `<pre>` element will get a `data-language` attribute with a
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/converters.ts:37]:
```suggestion
 * @param useLabels When `true`, the `<pre>` element will get a `data-language` attribute with a
```

File: packages/ckeditor5-code-block/src/converters.ts [hunk 11]
```diff
@@ -52,31 +63,33 @@ export function modelToViewCodeBlockInsertion( model, languageDefs, useLabels =
...
 
 		if ( !consumable.consume( data.item, 'insert' ) ) {
 			return;
 		}
 
-		const codeBlockLanguage = data.item.getAttribute( 'language' );
+		const codeBlockLanguage = data.item.getAttribute( 'language' ) as string;
 		const targetViewPosition = mapper.toViewPosition( model.createPositionBefore( data.item ) );
-		const preAttributes = {};
+		const preAttributes: any = {};
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/converters.ts:75]:
```suggestion
		const preAttributes: Record<string, string> = {};
```

File: packages/ckeditor5-code-block/src/converters.ts [hunk 12]
```diff
@@ -174,10 +195,8 @@ export function dataViewToModelCodeBlockInsertion( editingView, languageDefs ) {
 		for ( const className of viewChildClasses ) {
 			const language = classesToLanguages[ className ];
 
-			if ( language ) {
-				writer.setAttribute( 'language', language, codeBlock );
-				break;
-			}
+			writer.setAttribute( 'language', language, codeBlock );
+			break;
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/converters.ts:199]: Why was this hanged? What about other classes that are not used for language?

File: packages/ckeditor5-code-block/src/converters.ts [hunk 13]
```diff
@@ -228,7 +251,7 @@ export function dataViewToModelTextNewlinesInsertion() {
 		consumable.consume( data.viewItem );
 
 		const text = data.viewItem.data;
-		const textLines = text.split( '\n' ).map( data => writer.createText( data ) );
+		const textLines = text.split( '\n' ).map( ( data: any ) => writer.createText( data ) );
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/converters.ts:254]:
```suggestion
		const textLines = text.split( '\n' ).map( data => writer.createText( data ) );
```

File: packages/ckeditor5-code-block/src/indentcodeblockcommand.ts [hunk 14]
```diff
@@ -16,27 +15,23 @@ import {
...
 
-		/**
-		 * A sequence of characters added to the line when the command is executed.
-		 *
-		 * @readonly
-		 * @private
-		 * @member {String}
-		 */
-		this._indentSequence = editor.config.get( 'codeBlock.indentSequence' );
+		this._indentSequence = editor.config.get( 'codeBlock.indentSequence' ) as string;
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/indentcodeblockcommand.ts:28]:
```suggestion
		this._indentSequence = editor.config.get( 'codeBlock.indentSequence' )!;
```

This type should be inferred from augmented `EditorConfig`.

File: packages/ckeditor5-code-block/src/outdentcodeblockcommand.ts [hunk 15]
```diff
@@ -21,7 +22,9 @@ import {
  * @extends module:core/command~Command
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/outdentcodeblockcommand.ts:22]: This is not needed.

File: packages/ckeditor5-code-block/src/outdentcodeblockcommand.ts [hunk 16]
```diff
@@ -21,7 +22,9 @@ import {
...
  */
 export default class OutdentCodeBlockCommand extends Command {
-	constructor( editor ) {
+	private readonly _indentSequence: string;
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/outdentcodeblockcommand.ts:25]: Missing docs.

File: packages/ckeditor5-code-block/src/outdentcodeblockcommand.ts [hunk 17]
```diff
@@ -31,13 +34,13 @@ export default class OutdentCodeBlockCommand extends Command {
 		 * @private
 		 * @member {String}
 		 */
-		this._indentSequence = editor.config.get( 'codeBlock.indentSequence' );
+		this._indentSequence = editor.config.get( 'codeBlock.indentSequence' ) as string;
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/outdentcodeblockcommand.ts:37]:
```suggestion
		this._indentSequence = editor.config.get( 'codeBlock.indentSequence' )!;
```

Same as in the indent command above. 

Docs should be moved to property declaration.

File: packages/ckeditor5-code-block/src/utils.ts [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+ * **Note:** The position is located after the leading white spaces in the text node.
+ */
+export function getIndentOutdentPositions( model: Model ): Array<Position> {
+	const selection = model.document.selection;
+	const positions: Array<Position> = [];
+
+	// When the selection is collapsed, there's only one position we can indent or outdent.
+	if ( selection.isCollapsed ) {
+		positions.push( selection.anchor! );
+		return positions;
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/utils.ts:202]:
```suggestion
		return [ selection.anchor! ];
```

File: packages/ckeditor5-code-block/src/utils.ts [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+ * @param schema Model's schema.
+ * @param element The element to be checked.
+ * @returns Check result.
+ */
+export function canBeCodeBlock( schema: Schema, element: Element ): boolean {
+	if ( element.is( 'rootElement' ) || schema.isLimit( element ) ) {
+		return false;
+	}
+
+	return schema.checkChild( element.parent as Element, 'codeBlock' );
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/utils.ts:258]:
```suggestion
	return schema.checkChild( element.parent!, 'codeBlock' );
```

**Review by niegowski — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 20]
```diff
@@ -123,42 +133,54 @@ export default class CodeBlockEditing extends Plugin {
 
 		// Allow all list* attributes on `codeBlock` (integration with DocumentList).
 		// Disallow all attributes on $text inside `codeBlock`.
-		schema.addAttributeCheck( ( context, attributeName ) => {
-			const isDocumentListAttributeOnCodeBlock = context.endsWith( 'codeBlock' ) &&
+		schema.addAttributeCheck( ( context: any, attributeName ) => {
+			const isDocumentListAttributeOnCodeBlock = context.endsWith &&
+				context.endsWith( 'codeBlock' ) &&
```
[the remark this answers is not in the document]
Comment (niegowski) [reply packages/ckeditor5-code-block/src/codeblockediting.ts]: I already pushed a fix for this (in this branch).

File: packages/ckeditor5-code-block/src/converters.ts (hunk 12 above)
Comment (mremiszewski) [reply packages/ckeditor5-code-block/src/converters.ts:199]: Good point, i;m not sure what happened here, reverting changes.

File: packages/ckeditor5-code-block/src/codeblock.ts [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+	interface PluginsMap {
+		[ CodeBlock.pluginName ]: CodeBlock;
+	}
+	interface EditorConfig {
+
+		/**
+		 * The configuration of the {@link module:code-block/codeblock~CodeBlock} feature.
+		 *
+		 * Read more in {@link module:code-block/codeblock~CodeBlockConfig}.
+		 *
```
Comment (niegowski) [inline packages/ckeditor5-code-block/src/codeblock.ts:53]:
```suggestion
```

**Review by niegowski — APPROVED** [review]

File: packages/ckeditor5-code-block/src/codeblockui.ts (hunk 9 above)
Comment (mremiszewski) [reply packages/ckeditor5-code-block/src/codeblockui.ts:66]: We decided with @arkflpc to change it to `any`

**Pull request merged**